provider/parameter.go

```diff
@@ -389,37 +384,63 @@ func fixValidationResourceData(rawConfig cty.Value, validation interface{}) (int
 	return vArr, nil
 }
 
-func valueIsType(typ OptionType, value string, attrPath cty.Path) diag.Diagnostics {
+func valueIsType(typ OptionType, value string) error {
 	switch typ {
 	case OptionTypeNumber:
 		_, err := strconv.ParseFloat(value, 64)
 		if err != nil {
-			return diag.Errorf("%q is not a number", value)
+			return fmt.Errorf("%q is not a number", value)
 		}
 	case OptionTypeBoolean:
 		_, err := strconv.ParseBool(value)
 		if err != nil {
-			return diag.Errorf("%q is not a bool", value)
+			return fmt.Errorf("%q is not a bool", value)
 		}
 	case OptionTypeListString:
-		_, diags := valueIsListString(value, attrPath)
-		if diags.HasError() {
-			return diags
+		_, err := valueIsListString(value)
+		if err != nil {
+			return err
 		}
 	case OptionTypeString:
 		// Anything is a string!
 	default:
-		return diag.Errorf("invalid type %q", typ)
+		return fmt.Errorf("invalid type %q", typ)
 	}
 	return nil
 }
 
-func (v *Parameter) Valid() diag.Diagnostics {
+func (v *Parameter) Valid(input *string, mode ValidationMode) (string, diag.Diagnostics) {
+	if mode != ValidationModeDefault && mode != ValidationModeTemplateImport {
+		return "", diag.Diagnostics{
+			{
+				Severity: diag.Error,
+				Summary:  "Invalid validation mode",
+				Detail:   fmt.Sprintf("validation mode %q is not supported, use %q, or %q", mode, ValidationModeDefault, ValidationModeTemplateImport),
+			},
+		}
+	}
+
+	var err error
+	var optionType OptionType
+
+	value := input
+	if input == nil {
+		value = v.Default
+	}
+
+	// TODO: When empty values want to be rejected, uncomment this.
+	//   coder/coder should update to use template import mode first,
+	//   before this is uncommented.
+	//if value == nil && mode == ValidationModeDefault {
+	//	var empty string
+	//	value = &empty
+	//}
```

Comment:
This business wouldn't be necessary if the provider assumes the "looser" validation mode if unset. That is, there are 2 modes: WorkspaceCreate and TemplateImport, which in future we will explicitly set via environment variable. If unset, we assume TemplateImport, which has no breaking changes from current behavior.

Reply:
I removed this and did the early `return` instead as you suggested.

`TemplateImport` actually has more breaking changes than `create` in the end. `create` is backwards compatible because imported templates already exist and must still work.

Newly imported templates (once coderd switches), is going to require every `option` to also be valid. By doing this, the workspace create form will only present valid options to the user.


Eventually I would like to disallow an empty input value at `create` entirely.